keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -333,36 +336,169 @@ or later. The dual-stack feature requires no changes to this API.
+
+##### Creating a New Dual Stack Service
+
+Users can create dual stack services and control which up gets assigned to Service Primary IP as the following:
+1. Set `spec.ipFamilyPolicy: DualStack`. 
+2. If `spec.ipFamilyPolicy: nil` and `len(spec.ipFamilies) == 2` the resulting service will have a defaulted `spec.ipFamilyPolicy: DualStack`.
+3. If `spec.ipFamilies: nil` then it will be defaulted to the same order of ip families as set in `--services-cluster-ip-range`
+4. If `spec.ipFamilies: ['ip family']` it will be defaulted to `spec.ipFamilies: ['ip family', 'other ip family']`.
+5. If `spec.ipFamilies: ['ip family', 'ip family']` the order will be respected by the system and `spec.ClusterIP` will be allocated from the first entry in the array.
+
+> It is implicitly assumed that a Service with `spec.ipFamilyPolicy: nil` is always a single stack, however users can change this behavior and the order of ip families using admission controllers.
+
+The below is an example for the dual stack Service where the user only set  `spec.ipFamilyPolicy: DualStack` on a cluster that has `--services-cluster-ip-range=<ipc4 cidr>, <ipv6 cidr>`
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  type: ClusterIP
+  ipFamilyPolicy: DualStack # set by user
+  ipFamilies: # defaulted
+    - IPv4
+    - IPv6
+  clusterIP: 1.2.3.4  # allocated
+  clusterIPs: # allocated
+    - 1.2.3.4
+    - 2001::1
+  selector:
+    app: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+```
 
+The below Service is a dual stack service where the user is requesting a certain order of ip family assignment, irrespective of the `--services-cluster-ip-range` setting.
+ 
 ```yaml
 apiVersion: v1
 kind: Service
 metadata:
   name: my-service
 spec:
-  ipFamily: IPv6
+  type: ClusterIP
+  ipFamilyPolicy: DualStack # defaulted
+  ipFamilies: # set by user
+    - IPv6
+    - IPv4
+  clusterIP: 2001::1  # allocated
+  clusterIPs: # allocated
+    - 2001::1
+    - 1.2.3.4
   selector:
     app: MyApp
   ports:
     - protocol: TCP
       port: 80
       targetPort: 9376
+status:
+  ipFamilies:
+    - IPv4
+    - IPv6
 ```
 
-Dual-stack services (having both an IPv4 and IPv6 cluster IP) are currently out of scope of this proposal. Should there be use-cases where dual-stack Services are needed then we can revisit.
+##### Notes on NodePort Allocation
+
+NodePort reservation will be constant across both families. That mean a reservation of `NodePort: 12345` will happen across both families irrespective of the ip family of the service.
 
 #### Type Headless services
 
-\<TBD\>
+1. For Headless with selector, the value of `spec.ipFamilyPolicy` and `spec.ipFamilies` will drive the endpoint selection. Endpoint slice controller will select endpoints with ip families that match the value of these two flags.
```

Comment:
And EndpointController?

Reply:
Clarify that `ipFamilyPolicy` and `ipFamilies` are defaulted in the same way as for `ClusterIP` services

Reply:
So Tim implied above that he assumed headless services would look like:

    spec:
      clusterIP: None
      clusterIPs:
      - None

but I feel like it would be nicer to say that `clusterIPs` only ever contains actual IPs, and that for a headless service, `clusterIPs` is `nil` and it's only `clusterIP` that has the `None` value...

Reply:
If "None" is how you specify headless, I think it has to be consistent.  I don't like it, but we can't really add another `type` (well, we could but this still has to work)

Reply:
Well, the existing system is already inconsistent, by sometimes putting a non-IP-address value into a field that's supposed to be an IP address, and [this inconsistency caused us to screw up headless services in the original dual-stack implementation](https://github.com/kubernetes/kubernetes/pull/86895#issuecomment-577070772).

Saying "`clusterIP` mirrors `clusterIPs` except in the case of headless services" is less consistent in one way, but more consistent in another, since it means that `clusterIPs` only ever contains actual IP addresses. And it looks nicer...

Reply:
if/when we ever get rid of `clusterIP` (e.g. v2) then we have no representation left for headless.  I suppose at that point we could add a different type or a new field.

I guess I don't immediately see any problem with it.  @khenidak  ?

Reply:
I vote for carrying inconsistency wre `None` and `empty string` forward. I realize it is not as clean as @danwinship, but it is more consistent and functional.

Not just because it simplify the conversion/validation logic (which is already complex when you think of four interdependent fields `IPFamilyPolicy, IPFamilies, ClusterIPs, Type`. But it also simplifies the user documentation think of `you can have None || empty string in ClusterIP, but not in ClusterIPs in case of x, y,z types` vs `if it works on ClusterIP it will work for ClusterIPs[0]`. And it helps users their modifying existing code instead of `modify` code, users will only need to `add new code` for dual stack support.

As i am writing this I am modifying `apis/core/validation/validation_test.go` and i can tell the amount of effort needed by a user to modify for `spec.ClusterIPs= [] string{None}`is a lot less than `spec.ClusterIPs= nil in some ClusterIP == None || ClusterIP == None cases`.

The above opinion is a strong one. But i am not holding strongly to it. And i will change if the above is not convincing enough.